### PR TITLE
HD-3340 drain stream when receiving an error on a PUT

### PR DESCRIPTION
**Pull request by benoit-a**

### CODE REVIEW

File: src/hdcontroller.ts [hunk 1]
```diff
@@ -298,14 +298,16 @@ export class HDProxydClient {
     public put(stream: Stream, size: number, params: any, reqUids: string, callback: HDProxydClientPutCallback): void {
         const log = this.createLogger(reqUids);
         this._failover('POST', stream, size, '', 0, log, (err?: HDProxydError, response?: http.IncomingMessage) => {
+            if (response) {
+                response.resume();
+            }
```
Comment (vietnscality) [inline src/hdcontroller.ts:317]: I dont understand why moving the `response.resume()` here solves the issue.

Basically, if there is a response of the request, we force the stream to continue to flow so we can read the data? The `end` event of the stream will come later and trigger a close?

Comment (vietnscality) [reply src/hdcontroller.ts:317]: Maybe we need to check the err too?

```
if (err && response) {
    response.resume();
}

Comment (benoit-a) [reply src/hdcontroller.ts:317]: Good catch, it does not fix the issue.
It was intended when/if we have both an error and a response in the same call.

In the issue at hands, it is not sufficient.
I am considering letting this code (it is not harmful and is doing the right thing IMO), though

Comment (vietnscality) [reply src/hdcontroller.ts:317]: OK for keeping this code

**Review by vietnscality — APPROVED** [review]

**Review by ghost — COMMENTED** [review]
Changes LGTM but given that it's a bugfix ficket, I would prefer that we also create some new unit tests to ensure that the fix is enough, and that we don't reproduce this problem again in the future

**Comment by vietnscality** [comment]
> Changes LGTM but given that it's a bugfix ficket, I would prefer that we also create some new unit tests to ensure that the fix is enough, and that we don't reproduce this problem again in the future

@williamlardier Maybe you did not see this comment https://github.com/scality/hdclient/pull/659/files#r1356780040

**Comment by ghost** [comment]
Hello @vietnscality , yes I saw the comment, mine is more about adding some unit tests to reproduce the initial problem and making sure we won't have any regression for this behaviour in the future

**Comment by vietnscality** [comment]
> Hello @vietnscality , yes I saw the comment, mine is more about adding some unit tests to reproduce the initial problem and making sure we won't have any regression for this behaviour in the future

Yes, i agree, it is a good point. The pb is AWAIU since the PR does not fix the issue, having a test will fail the CI. Ideally the unit test should go together with the fix. Or maybe I missed something?

**Comment by ghost** [comment]
@vietnscality we can `.skip`  it if needed, till we have a proper fix; but what I had in mind was to test the current code, i.e. consider that we resume the readable stream even if `err`  is not defined (and update it later with the real fix), make sure there is no side effects upstream (in cloudserver) with the current code, and don't add too much technical debt 

It's not a blocking comment anyway :smile:

**Comment by vietnscality** [comment]
> @vietnscality we can `.skip` it if needed, till we have a proper fix; but what I had in mind was to test the current code, i.e. consider that we resume the readable stream even if `err` is not defined (and update it later with the real fix), make sure there is no side effects upstream (in cloudserver) with the current code, and don't add too much technical debt
> 
> It's not a blocking comment anyway 😄

OK, fair enough, I'll let the author @benoit-a decide what is appropriate for him

**Comment by benoit-a** [comment]
> OK, fair enough, I'll let the author @benoit-a decide what is appropriate for him

Good point for the test part, I will how I can do to detect the behaviour on a unit test

Regarding the rest of the discussion, I think there is confusion ; to me, the real bug is here, hdclient MUST close the session and for that, it has to drain the stream. It should do that every time there is a response

The changes in hdcontroller are merely "good practices" that could have helped slightly but the client part (cloudserver) would still exhibit issues (possibly accumulating sockets in CLOSE WAIT or file descriptor)

**Review by ghost — APPROVED** [review]
LGTM!

**Pull request merged**